File: server/src/text_preprocessor.py
"""Text preprocessing for TTS - cleaning, normalization, and chunking."""

import re
import unicodedata
from typing import List

from num2words import num2words

from .config import settings
from .logging_config import get_logger, preview_text

logger = get_logger(__name__)


class TextPreprocessor:
    """Preprocess text for optimal TTS quality."""

    # Common abbreviations to expand
    ABBREVIATIONS = {
        r'\bDr\.': 'Doctor',
        r'\bMr\.': 'Mister',
        r'\bMrs\.': 'Missus',
        r'\bMs\.': 'Miss',
        r'\bProf\.': 'Professor',
        r'\bSr\.': 'Senior',
        r'\bJr\.': 'Junior',
        r'\bInc\.': 'Incorporated',
        r'\bLtd\.': 'Limited',
        r'\bCo\.': 'Company',
        r'\bCorp\.': 'Corporation',
        r'\bAve\.': 'Avenue',
        r'\bSt\.': 'Street',
        r'\bRd\.': 'Road',
        r'\bBlvd\.': 'Boulevard',
        r'\betc\.': 'et cetera',
        r'\be\.g\.': 'for example',
        r'\bi\.e\.': 'that is',
        r'\bvs\.': 'versus',
    }

    # Emoji regex pattern (covers most common emoji ranges)
    EMOJI_PATTERN = re.compile(
        "["
        "\U0001F600-\U0001F64F"  # emoticons
        "\U0001F300-\U0001F5FF"  # symbols & pictographs
        "\U0001F680-\U0001F6FF"  # transport & map
        "\U0001F1E0-\U0001F1FF"  # flags
        "\U00002702-\U000027B0"  # dingbats
        "\U000024C2-\U0001F251"  # misc
        "\U0001F900-\U0001F9FF"  # supplemental symbols
        "\U0001FA00-\U0001FA6F"  # chess symbols
        "\U0001FA70-\U0001FAFF"  # symbols extended
        "\U00002600-\U000026FF"  # misc symbols
        "\U00002700-\U000027BF"  # dingbats
        "]+",
        flags=re.UNICODE
    )

    def __init__(self, max_chunk_tokens: int = None):
        """
        Initialize the text preprocessor.

        Args:
            max_chunk_tokens: Maximum tokens per chunk (default from settings)
        """
        self.max_chunk_tokens = max_chunk_tokens or settings.MAX_CHUNK_TOKENS

    def normalize(self, text: str) -> str:
        """
        Normalize text for TTS processing.

        Full pipeline:
        1. Unicode normalization
        2. Strip emojis
        3. Strip markdown formatting
        4. Sanitize special characters
        5. Clean PDF artifacts
        6. Expand abbreviations
        7. Convert numbers to words
        8. Final whitespace cleanup

        Args:
            text: Raw input text

        Returns:
            Normalized text ready for TTS
        """
        original_len = len(text)
        original_newlines = text.count('\n')
        original_paragraphs = text.count('\n\n') + 1

        logger.debug(f"Normalize input: {original_len} chars, {original_newlines} newlines, ~{original_paragraphs} paragraphs")
        logger.debug(f"Input preview: {preview_text(text, 300)}")

        # 1. Normalize Unicode (NFKC handles ligatures, special chars)
        text = unicodedata.normalize('NFKC', text)

        # 2. Strip emojis (before any text processing)
        text = self._strip_emojis(text)

        # 3. Strip markdown formatting
        text = self._strip_markdown(text)

        # 4. Sanitize special characters (URLs, emails, symbols)
        text = self._sanitize_special_chars(text)

        # 5. Remove PDF artifacts (page numbers, excessive whitespace)
        text = re.sub(r'\n\s*\d+\s*\n', '\n', text)  # Page numbers
        text = re.sub(r'\s{3,}', ' ', text)  # Multiple spaces
        text = re.sub(r'\n{3,}', '\n\n', text)  # Excessive newlines

        # Remove hyphenation at line breaks
        text = re.sub(r'-\n', '', text)

        # 6. Expand common abbreviations
        for pattern, replacement in self.ABBREVIATIONS.items():
            text = re.sub(pattern, replacement, text, flags=re.IGNORECASE)

        # 7. Convert numbers to words (for better TTS pronunciation)
        text = self._convert_numbers_to_words(text)

        # 8. Clean up whitespace while PRESERVING paragraph breaks
        # First normalize line endings
        text = text.replace('\r\n', '\n').replace('\r', '\n')

        # Count paragraphs before whitespace normalization
        pre_whitespace_paragraphs = text.count('\n\n') + 1

        # Collapse multiple spaces (but not newlines) to single space
        text = re.sub(r'[^\S\n]+', ' ', text)
        # Normalize paragraph breaks: 2+ newlines become double newline
        text = re.sub(r'\n{2,}', '\n\n', text)

        # Handle single newlines intelligently:
        # - If line ends with sentence punctuation (.!?) and next line starts with
        #   uppercase or number, treat as paragraph break (convert to double newline)
        # - Otherwise convert to space (soft wrap)
        def handle_single_newline(match):
            before = match.group(1)  # Character before newline
            after = match.group(2)   # Character after newline
            # Sentence end followed by new sentence = paragraph break
            if before in '.!?:' and (after.isupper() or after.isdigit()):
                return before + '\n\n' + after
            # Otherwise it's a soft wrap
            return before + ' ' + after

        # Match: char + single newline (not preceded/followed by newline) + char
        text = re.sub(r'([^\n])\n(?!\n)([^\n])', handle_single_newline, text)

        # Clean up spaces around paragraph breaks
        text = re.sub(r' *\n\n *', '\n\n', text)
        # Strip leading/trailing whitespace from each line
        text = '\n\n'.join(line.strip() for line in text.split('\n\n'))

        final_len = len(text)
        final_newlines = text.count('\n')
        final_paragraphs = text.count('\n\n') + 1

        logger.info(f"Normalize: {original_len} -> {final_len} chars, {original_newlines} -> {final_newlines} newlines, {pre_whitespace_paragraphs} -> {final_paragraphs} paragraphs")
        logger.debug(f"Output preview: {preview_text(text, 300)}")

        return text.strip()

    def _strip_emojis(self, text: str) -> str:
        """
        Remove emojis to prevent TTS issues.

        Args:
            text: Input text potentially containing emojis

        Returns:
            Text with emojis removed
        """
        return self.EMOJI_PATTERN.sub(' ', text)

    def _strip_markdown(self, text: str) -> str:
        """
        Strip markdown formatting for clean TTS output.

        Handles: headers, bold, italic, links, code, horizontal rules,
        list markers, and blockquotes.

        Args:
            text: Input text potentially containing markdown

        Returns:
            Plain text without markdown formatting
        """
        # Code blocks: ```lang\ncode\n``` → code (do this first)
        text = re.sub(r'```[^\n]*\n(.*?)```', r'\1', text, flags=re.DOTALL)

        # Inline code: `code` → code
        text = re.sub(r'`([^`]+)`', r'\1', text)

        # Headers: ## Title → Title
        text = re.sub(r'^#{1,6}\s+', '', text, flags=re.MULTILINE)

        # Bold/italic: **text** or *text* or __text__ or _text_ → text
        text = re.sub(r'\*\*([^*]+)\*\*', r'\1', text)
        text = re.sub(r'\*([^*]+)\*', r'\1', text)
        text = re.sub(r'__([^_]+)__', r'\1', text)
        text = re.sub(r'_([^_]+)_', r'\1', text)

        # Links: [text](url) → text
        text = re.sub(r'\[([^\]]+)\]\([^)]+\)', r'\1', text)

        # Images: ![alt](url) → alt
        text = re.sub(r'!\[([^\]]*)\]\([^)]+\)', r'\1', text)

        # Horizontal rules: --- or *** or ___ → (remove)
        text = re.sub(r'^[-*_]{3,}\s*$', '', text, flags=re.MULTILINE)

        # List markers: - item or * item or + item → item
        text = re.sub(r'^\s*[-*+]\s+', '', text, flags=re.MULTILINE)

        # Numbered lists: 1. item → item
        text = re.sub(r'^\s*\d+\.\s+', '', text, flags=re.MULTILINE)

        # Blockquotes: > text → text
        text = re.sub(r'^>\s*', '', text, flags=re.MULTILINE)

        return text

    def _sanitize_special_chars(self, text: str) -> str:
        """
        Replace or remove special characters that shouldn't be vocalized.

        Handles: URLs, emails, file paths, hashtags, mentions,
        and various symbols.

        Args:
            text: Input text with special characters

        Returns:
            Text with special characters handled appropriately
        """
        # URLs → remove (handles URLs in parentheses, brackets, quotes)
        # Match http/https URLs including those ending with punctuation
        text = re.sub(r'https?://[^\s<>"\'()\[\]{}]+[^\s<>"\'()\[\]{},.:;!?]', ' ', text)
        # Also catch URLs that might be wrapped in parentheses: (https://...)
        text = re.sub(r'\(https?://[^)]+\)', ' ', text)
        # And in brackets: [https://...]
        text = re.sub(r'\[https?://[^\]]+\]', ' ', text)

        # Remove any remaining URL-like patterns (www.example.com)
        text = re.sub(r'www\.[^\s<>"\'()\[\]{}]+', ' ', text)

        # Email addresses → remove
        text = re.sub(r'[\w.+-]+@[\w.-]+\.\w{2,}', ' ', text)

        # File paths: /usr/bin/file or C:\path\file → remove
        text = re.sub(r'(?:[A-Za-z]:)?[/\\][\w./\\-]+', ' ', text)

        # Reference-style patterns often from copy-paste:
        # [1], [2], [citation], (source), (ref 1), etc.
        text = re.sub(r'\[\d+\]', ' ', text)  # [1], [23]
        text = re.sub(r'\(\d+\)', ' ', text)  # (1), (23)

        # Parenthetical metadata: (PDF), (link), (source: ...), (via ...), (accessed ...)
        text = re.sub(r'\((?:PDF|Link|Source|Via|Accessed|Retrieved|See|Cf\.?)[^)]*\)', ' ', text, flags=re.IGNORECASE)

        # Empty parentheses or brackets left after URL removal
        text = re.sub(r'\(\s*\)', ' ', text)
        text = re.sub(r'\[\s*\]', ' ', text)

        # Hashtags: #topic → topic (keep word, remove #)
        text = re.sub(r'#(\w+)', r'\1', text)

        # At-mentions: @username → username
        text = re.sub(r'@(\w+)', r'\1', text)

        # HTML/XML tags: <tag> → remove
        text = re.sub(r'<[^>]+>', ' ', text)

        # Ampersand → and
        text = re.sub(r'\s*&\s*', ' and ', text)

        # Plus sign between words → plus
        text = re.sub(r'\s+\+\s+', ' plus ', text)

        # Percent → percent
        text = re.sub(r'(\d+)\s*%', r'\1 percent', text)

        # Degree symbol → degrees
        text = re.sub(r'(\d+)\s*°([CF]?)', lambda m: f"{m.group(1)} degrees {m.group(2)}", text)

        # Copyright/trademark symbols
        text = re.sub(r'©', ' copyright ', text)
        text = re.sub(r'®', ' registered ', text)
        text = re.sub(r'™', ' trademark ', text)

        # Standalone special chars that shouldn't be spoken
        # Remove: / \ | ~ ^ ` < > { } [ ]
        text = re.sub(r'[/\\|~^`<>{}\[\]]', ' ', text)

        # Remove remaining standalone asterisks (not part of words)
        text = re.sub(r'\s\*+\s', ' ', text)
        text = re.sub(r'^\*+\s', '', text, flags=re.MULTILINE)

        return text

    def _convert_numbers_to_words(self, text: str) -> str:
        """
        Convert standalone numbers to words for better TTS.

        Handles both integers and decimal numbers.

        Args:
            text: Input text with numbers

        Returns:
            Text with numbers converted to words
        """
        def replace_decimal(match):
            """Convert decimal numbers: 3.14 → three point one four"""
            try:
                parts = match.group(0).split('.')
                whole = num2words(int(parts[0])) if parts[0] else 'zero'
                # Convert each decimal digit to words
                decimal_digits = ' '.join(num2words(int(d)) for d in parts[1])
                return f"{whole} point {decimal_digits}"
            except (ValueError, OverflowError):
                return match.group(0)

        def replace_integer(match):
            """Convert integers: 42 → forty-two"""
            try:
                num = match.group(0)
                # Only convert reasonable numbers (not years, IDs, etc.)
                if len(num) <= 4:
                    return num2words(int(num))
                return num
            except (ValueError, OverflowError):
                return match.group(0)

        # First handle decimals: 3.14, 0.5, etc.
        text = re.sub(r'\b\d+\.\d+\b', replace_decimal, text)

        # Then handle remaining standalone integers
        text = re.sub(r'\b\d{1,4}\b', replace_integer, text)

        return text

    def chunk_text(self, text: str) -> List[dict]:
        """
        Split text into optimal chunks for TTS processing.

        Chunks are split at sentence boundaries and target max_chunk_tokens.
        This ensures natural pauses and stays within Kokoro's 510 token limit.
        Paragraph breaks are preserved in the chunk metadata.

        Args:
            text: Preprocessed text to chunk

        Returns:
            List of chunk dicts: [{"text": str, "starts_paragraph": bool}, ...]
        """
        # First split into paragraphs to preserve structure
        paragraphs = text.split('\n\n')

        chunks = []

        for para_idx, paragraph in enumerate(paragraphs):
            if not paragraph.strip():
                continue

            # Split paragraph into sentences
            sentences = re.split(r'(?<=[.!?])\s+', paragraph.strip())

            current_chunk = []
            current_length = 0
            is_first_in_para = True

            for sentence in sentences:
                if not sentence.strip():
                    continue

                # Rough token estimate: ~4 characters per token
                sentence_tokens = len(sentence) // 4

                # If adding this sentence exceeds max, save current chunk
                if current_length + sentence_tokens > self.max_chunk_tokens and current_chunk:
                    chunks.append({
                        "text": ' '.join(current_chunk),
                        "starts_paragraph": is_first_in_para
                    })
                    current_chunk = [sentence]
                    current_length = sentence_tokens
                    is_first_in_para = False
                else:
                    current_chunk.append(sentence)
                    current_length += sentence_tokens

            # Add remaining sentences from this paragraph
            if current_chunk:
                chunks.append({
                    "text": ' '.join(current_chunk),
                    "starts_paragraph": is_first_in_para
                })

        return chunks

    def process(self, text: str) -> List[dict]:
        """
        Full preprocessing pipeline: normalize and chunk text.

        Args:
            text: Raw input text

        Returns:
            List of chunk dicts: [{"text": str, "starts_paragraph": bool}, ...]
        """
        logger.info(f"Processing text: {len(text)} chars input")
        normalized = self.normalize(text)
        chunks = self.chunk_text(normalized)
        logger.info(f"Processing complete: {len(chunks)} chunks created")
        return chunks

File: android/app/src/main/java/com/openmobiletts/app/TextPreprocessor.kt
package com.openmobiletts.app

import java.text.Normalizer

/**
 * Text preprocessing for TTS — normalization, sanitization, and chunking.
 *
 * Ports the desktop Python TextPreprocessor to Android to ensure high-quality
 * TTS output. Raw text from document extraction or user input goes through
 * an 8-step normalization pipeline before being chunked into TTS-ready segments.
 *
 * Pipeline:
 *   1. Unicode NFKC normalization (ligatures, smart quotes)
 *   2. Emoji removal
 *   3. Markdown formatting removal
 *   4. Special character sanitization (URLs, emails, symbols)
 *   5. PDF artifact cleanup (page numbers, hyphenated line breaks)
 *   6. Abbreviation expansion (Dr. → Doctor, etc.)
 *   7. Number-to-words conversion (42 → forty-two)
 *   8. Whitespace normalization (preserving paragraph breaks)
 *
 * Then chunks the normalized text at paragraph and sentence boundaries
 * with a token budget that stays safely under Kokoro's ~510 token limit.
 */
object TextPreprocessor {

    private const val TAG = "TextPreprocessor"

    // Chunking budget: Kokoro's token limit is ~510.
    // At ~4 chars/token, TARGET_TOKENS=250 ≈ 1000 chars, MAX_TOKENS=400 ≈ 1600 chars.
    private const val TARGET_TOKENS = 250
    private const val MAX_TOKENS = 400
    private const val CHARS_PER_TOKEN = 4

    /** A chunk of text ready for TTS generation. */
    data class TextChunk(val text: String, val startsParagraph: Boolean)

    // Common abbreviations → expansions (case-insensitive regex)
    private val ABBREVIATION_REGEXES: List<Pair<Regex, String>> = listOf(
        "\\bDr\\." to "Doctor",
        "\\bMr\\." to "Mister",
        "\\bMrs\\." to "Missus",
        "\\bMs\\." to "Miss",
        "\\bProf\\." to "Professor",
        "\\bSr\\." to "Senior",
        "\\bJr\\." to "Junior",
        "\\bInc\\." to "Incorporated",
        "\\bLtd\\." to "Limited",
        "\\bCo\\." to "Company",
        "\\bCorp\\." to "Corporation",
        "\\bAve\\." to "Avenue",
        "\\bSt\\." to "Street",
        "\\bRd\\." to "Road",
        "\\bBlvd\\." to "Boulevard",
        "\\betc\\." to "et cetera",
        "\\be\\.g\\." to "for example",
        "\\bi\\.e\\." to "that is",
        "\\bvs\\." to "versus",
    ).map { (pattern, replacement) ->
        Regex(pattern, RegexOption.IGNORE_CASE) to replacement
    }

    // Number words lookup tables
    private val ONES = arrayOf(
        "zero", "one", "two", "three", "four", "five", "six", "seven",
        "eight", "nine", "ten", "eleven", "twelve", "thirteen", "fourteen",
        "fifteen", "sixteen", "seventeen", "eighteen", "nineteen",
    )
    private val TENS = arrayOf(
        "", "", "twenty", "thirty", "forty", "fifty",
        "sixty", "seventy", "eighty", "ninety",
    )

    // ---------- Public API ----------

    /**
     * Full preprocessing pipeline: normalize text and chunk for TTS.
     * @param text Raw input text (from user or document extraction)
     * @return List of text chunks ready for TTS generation
     */
    fun process(text: String): List<TextChunk> {
        val inputLen = text.length
        AppLog.i(TAG, "Processing $inputLen chars")
        val normalized = normalize(text)
        val chunks = chunkText(normalized)
        AppLog.i(TAG, "Processed: $inputLen → ${normalized.length} chars, ${chunks.size} chunks")
        return chunks
    }

    /**
     * Normalize text through the 8-step pipeline.
     */
    fun normalize(text: String): String {
        var t = text

        // 1. Unicode NFKC (collapses ligatures, smart quotes, half-width chars)
        t = Normalizer.normalize(t, Normalizer.Form.NFKC)

        // 2. Strip emojis
        t = stripEmojis(t)

        // 3. Strip markdown formatting
        t = stripMarkdown(t)

        // 4. Sanitize special characters (URLs, emails, symbols)
        t = sanitizeSpecialChars(t)

        // 5. Clean PDF artifacts (page numbers, hyphenation, excess whitespace)
        t = cleanPdfArtifacts(t)

        // 6. Expand abbreviations (Dr. → Doctor, etc.)
        t = expandAbbreviations(t)

        // 7. Convert numbers to words (42 → forty-two)
        t = convertNumbers(t)

        // 8. Clean whitespace while preserving paragraph breaks
        t = cleanWhitespace(t)

        return t.trim()
    }

    /**
     * Chunk normalized text at paragraph and sentence boundaries.
     * Each chunk stays within the token budget for Kokoro TTS.
     */
    fun chunkText(text: String): List<TextChunk> {
        val paragraphs = text.split(Regex("\n{2,}"))
            .map { it.trim() }
            .filter { it.isNotEmpty() }

        val result = mutableListOf<TextChunk>()

        for (paragraph in paragraphs) {
            val sentences = splitSentences(paragraph)
            val grouped = groupSentences(sentences)

            for ((i, chunkText) in grouped.withIndex()) {
                result.add(TextChunk(text = chunkText, startsParagraph = i == 0))
            }
        }

        // Fallback: if no chunks produced, use the whole text
        if (result.isEmpty() && text.isNotBlank()) {
            result.add(TextChunk(text = text.trim(), startsParagraph = true))
        }

        return result
    }

    // ---------- Step 2: Emoji removal ----------

    private fun stripEmojis(text: String): String {
        val sb = StringBuilder(text.length)
        var i = 0
        while (i < text.length) {
            val cp = text.codePointAt(i)
            val charCount = Character.charCount(cp)

            val isEmoji = cp in 0x1F600..0x1F64F ||  // Emoticons
                    cp in 0x1F300..0x1F5FF ||  // Misc Symbols and Pictographs
                    cp in 0x1F680..0x1F6FF ||  // Transport and Map
                    cp in 0x1F1E0..0x1F1FF ||  // Flags
                    cp in 0x1F900..0x1F9FF ||  // Supplemental Symbols
                    cp in 0x1FA00..0x1FA6F ||  // Chess Symbols
                    cp in 0x1FA70..0x1FAFF ||  // Extended Symbols
                    cp in 0x2600..0x26FF ||    // Misc Symbols
                    cp in 0x2702..0x27B0 ||    // Dingbats
                    cp in 0x2700..0x27BF ||    // Dingbats extended
                    cp in 0xFE00..0xFE0F ||    // Variation Selectors
                    cp == 0x200D ||             // Zero Width Joiner
                    cp == 0x20E3               // Combining Enclosing Keycap

            if (!isEmoji) {
                sb.appendCodePoint(cp)
            } else {
                sb.append(' ')
            }
            i += charCount
        }
        return sb.toString()
    }

    // ---------- Step 3: Markdown removal ----------

    private fun stripMarkdown(text: String): String {
        var t = text
        // Code blocks (fenced) — keep content, strip delimiters
        t = t.replace(Regex("```[^\\n]*\\n([\\s\\S]*?)```")) { it.groupValues[1] }
        // Inline code → content
        t = t.replace(Regex("`([^`]+)`")) { it.groupValues[1] }
        // Images ![alt](url) → alt
        t = t.replace(Regex("!\\[([^\\]]*)\\]\\([^)]*\\)"), "$1")
        // Links [text](url) → text
        t = t.replace(Regex("\\[([^\\]]*)\\]\\([^)]*\\)"), "$1")
        // HTML tags
        t = t.replace(Regex("<[^>]+>"), "")
        // Headers
        t = t.replace(Regex("(?m)^#{1,6}\\s+"), "")
        // Bold/italic (symmetric delimiters only — process bold before italic)
        t = t.replace(Regex("\\*\\*([^*]+)\\*\\*"), "$1")
        t = t.replace(Regex("\\*([^*]+)\\*"), "$1")
        t = t.replace(Regex("__([^_]+)__"), "$1")
        t = t.replace(Regex("_([^_]+)_"), "$1")
        // Strikethrough
        t = t.replace(Regex("~~([^~]+)~~"), "$1")
        // Blockquotes
        t = t.replace(Regex("(?m)^>\\s?"), "")
        // Horizontal rules
        t = t.replace(Regex("(?m)^[-*_]{3,}\\s*$"), "")
        // List markers (bullet and numbered)
        t = t.replace(Regex("(?m)^\\s*[-*+]\\s+"), "")
        t = t.replace(Regex("(?m)^\\s*\\d+\\.\\s+"), "")
        return t
    }

    // ---------- Step 4: Special character sanitization ----------

    private fun sanitizeSpecialChars(text: String): String {
        var t = text
        // URLs (http/https)
        t = t.replace(Regex("https?://[^\\s<>\"'()\\[\\]{}]+[^\\s<>\"'()\\[\\]{},.:;!?]"), " ")
        t = t.replace(Regex("\\(https?://[^)]+\\)"), " ")
        t = t.replace(Regex("\\[https?://[^\\]]+\\]"), " ")
        t = t.replace(Regex("www\\.[^\\s<>\"'()\\[\\]{}]+"), " ")
        // Email addresses
        t = t.replace(Regex("[\\w.+-]+@[\\w.-]+\\.\\w{2,}"), " ")
        // File paths
        t = t.replace(Regex("(?:[A-Za-z]:)?[/\\\\][\\w./\\\\-]+"), " ")
        // Reference numbers [1], (1)
        t = t.replace(Regex("\\[\\d+\\]"), " ")
        t = t.replace(Regex("\\(\\d+\\)"), " ")
        // Parenthetical metadata: (PDF), (Source: ...), etc.
        t = t.replace(
            Regex("\\((?:PDF|Link|Source|Via|Accessed|Retrieved|See|Cf\\.?)[^)]*\\)", RegexOption.IGNORE_CASE),
            " ",
        )
        // Empty parens/brackets left after URL removal
        t = t.replace(Regex("\\(\\s*\\)"), " ")
        t = t.replace(Regex("\\[\\s*\\]"), " ")
        // Hashtags → word
        t = t.replace(Regex("#(\\w+)"), "$1")
        // @mentions → word
        t = t.replace(Regex("@(\\w+)"), "$1")
        // Ampersand → and
        t = t.replace(Regex("\\s*&\\s*"), " and ")
        // Plus sign between words → plus
        t = t.replace(Regex("\\s+\\+\\s+"), " plus ")
        // Percent → percent
        t = t.replace(Regex("(\\d+)\\s*%"), "$1 percent")
        // Degree → degrees
        t = t.replace(Regex("(\\d+)\\s*°([CF]?)")) { m ->
            "${m.groupValues[1]} degrees ${m.groupValues[2]}"
        }
        // Copyright/trademark symbols
        t = t.replace("©", " copyright ")
        t = t.replace("®", " registered ")
        t = t.replace("™", " trademark ")
        // Standalone special chars that shouldn't be spoken
        t = t.replace(Regex("[/\\\\|~^`<>{}\\[\\]]"), " ")
        // Standalone asterisks
        t = t.replace(Regex("\\s\\*+\\s"), " ")
        t = t.replace(Regex("(?m)^\\*+\\s"), "")
        return t
    }

    // ---------- Step 5: PDF artifact cleanup ----------

    private fun cleanPdfArtifacts(text: String): String {
        var t = text
        // Page numbers on their own line: \n 42 \n
        t = t.replace(Regex("\n\\s*\\d+\\s*\n"), "\n")
        // Multiple horizontal spaces (not newlines — preserve paragraph breaks)
        t = t.replace(Regex("[^\\S\\n]{3,}"), " ")
        // Excessive newlines
        t = t.replace(Regex("\n{3,}"), "\n\n")
        // Hyphenated line breaks from PDF column wrapping
        t = t.replace(Regex("-\n"), "")
        return t
    }

    // ---------- Step 6: Abbreviation expansion ----------

    private fun expandAbbreviations(text: String): String {
        var t = text
        for ((regex, replacement) in ABBREVIATION_REGEXES) {
            t = t.replace(regex, replacement)
        }
        return t
    }

    // ---------- Step 7: Number-to-words conversion ----------

    private fun convertNumbers(text: String): String {
        // Decimals first: 3.14 → three point one four
        var t = text.replace(Regex("\\b(\\d+)\\.(\\d+)\\b")) { m ->
            decimalToWords(m.groupValues[1], m.groupValues[2])
        }
        // Then standalone integers (1-4 digits only, matching Python behavior)
        t = t.replace(Regex("\\b\\d{1,4}\\b")) { m ->
            try {
                intToWords(m.value.toInt())
            } catch (_: NumberFormatException) {
                m.value
            }
        }
        return t
    }

    private fun decimalToWords(intPart: String, decPart: String): String {
        val whole = try {
            intToWords(intPart.toInt())
        } catch (_: Exception) {
            intPart
        }
        val digits = decPart.map { c ->
            try {
                ONES[c.digitToInt()]
            } catch (_: Exception) {
                c.toString()
            }
        }.joinToString(" ")
        return "$whole point $digits"
    }

    /** Convert integer 0–9999 to English words. */
    private fun intToWords(n: Int): String {
        if (n < 0) return "negative ${intToWords(-n)}"
        return when {
            n < 20 -> ONES[n]
            n < 100 -> {
                val t = TENS[n / 10]
                if (n % 10 > 0) "$t-${ONES[n % 10]}" else t
            }
            n < 1000 -> {
                val rest = n % 100
                if (rest > 0) "${ONES[n / 100]} hundred and ${intToWords(rest)}"
                else "${ONES[n / 100]} hundred"
            }
            n < 10000 -> {
                val rest = n % 1000
                if (rest > 0) "${intToWords(n / 1000)} thousand ${intToWords(rest)}"
                else "${intToWords(n / 1000)} thousand"
            }
            else -> n.toString()
        }
    }

    // ---------- Step 8: Whitespace cleanup ----------

    private fun cleanWhitespace(text: String): String {
        var t = text
        // Normalize line endings
        t = t.replace("\r\n", "\n").replace("\r", "\n")
        // Collapse horizontal whitespace (not newlines)
        t = t.replace(Regex("[^\\S\n]+"), " ")
        // Normalize paragraph breaks
        t = t.replace(Regex("\n{2,}"), "\n\n")
        // Handle single newlines: sentence end + uppercase/digit = paragraph break, else space.
        // Loop until stable because each match consumes two boundary characters,
        // so consecutive single-newlines require multiple passes.
        var prev = ""
        while (prev != t) {
            prev = t
            t = t.replace(Regex("([^\n])\n(?!\n)([^\n])")) { m ->
                val before = m.groupValues[1]
                val after = m.groupValues[2]
                if (before.last() in ".!?:" && (after.first().isUpperCase() || after.first().isDigit())) {
                    "$before\n\n$after"
                } else {
                    "$before $after"
                }
            }
        }
        // Clean spaces around paragraph breaks and trim each paragraph
        t = t.replace(Regex(" *\n\n *"), "\n\n")
        t = t.split("\n\n").joinToString("\n\n") { it.trim() }
        return t
    }

    // ---------- Chunking helpers ----------

    /** Split text at sentence boundaries. */
    private fun splitSentences(text: String): List<String> {
        // After abbreviation expansion, most periods are true sentence ends.
        // Split at .!? followed by whitespace.
        return text.split(Regex("(?<=[.!?])\\s+"))
            .map { it.trim() }
            .filter { it.isNotEmpty() }
    }

    /** Group sentences into chunks within the token budget. */
    private fun groupSentences(sentences: List<String>): List<String> {
        val targetChars = TARGET_TOKENS * CHARS_PER_TOKEN  // ~1000
        val maxChars = MAX_TOKENS * CHARS_PER_TOKEN        // ~1600

        val result = mutableListOf<String>()
        val current = StringBuilder()

        for (sentence in sentences) {
            // Split very long sentences at clause boundaries first
            val parts = if (sentence.length > maxChars) {
                splitLongText(sentence, maxChars)
            } else {
                listOf(sentence)
            }

            for (part in parts) {
                if (current.isEmpty()) {
                    current.append(part)
                } else if (current.length + part.length + 1 <= targetChars) {
                    current.append(" ").append(part)
                } else {
                    result.add(current.toString())
                    current.clear()
                    current.append(part)
                }
            }
        }

        if (current.isNotEmpty()) {
            result.add(current.toString())
        }

        return result
    }

    /** Split text exceeding maxLength at clause or word boundaries. */
    private fun splitLongText(text: String, maxLength: Int): List<String> {
        if (text.length <= maxLength) return listOf(text)

        val result = mutableListOf<String>()
        var remaining = text

        while (remaining.length > maxLength) {
            var breakAt = -1

            // Prefer clause-level breaks (comma, semicolon, colon, em-dash)
            for (i in maxLength downTo maxLength / 2) {
                if (i < remaining.length && remaining[i] in ",;:\u2014\u2013") {
                    breakAt = i + 1
                    break
                }
            }

            // Fall back to word boundary
            if (breakAt == -1) {
                for (i in maxLength downTo maxLength / 2) {
                    if (i < remaining.length && remaining[i] == ' ') {
                        breakAt = i
                        break
                    }
                }
            }

            // Hard break as last resort
            if (breakAt == -1) breakAt = maxLength

            val chunk = remaining.substring(0, breakAt).trim()
            if (chunk.isNotEmpty()) result.add(chunk)
            val newRemaining = remaining.substring(breakAt).trim()
            if (newRemaining.length >= remaining.length) break // safety: ensure forward progress
            remaining = newRemaining
        }

        if (remaining.isNotEmpty()) result.add(remaining)

        return result
    }
}
